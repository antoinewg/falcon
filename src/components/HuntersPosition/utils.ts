// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const getErrorMessage = (huntersPosition: any, planets: string[]): string | undefined => {
  let errorMessage

  if (typeof huntersPosition !== 'object' || !huntersPosition) return 'The object is malformed'
  if (!('countdown' in huntersPosition)) return "The object should have the key 'countdown'"
  if (!('bounty_hunters' in huntersPosition))
    return "The object should have the key 'bounty_hunters'"

  const hunters = 'bounty_hunters' in huntersPosition ? huntersPosition.bounty_hunters : []
  if (!Array.isArray(hunters)) return "'bounty_hunters' should be an array"
  if (hunters.some(({ planet }) => !planets.includes(planet))) return 'Some planets are incorrect'

  return errorMessage
}
