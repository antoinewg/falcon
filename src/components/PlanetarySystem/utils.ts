import { Route, Link, Network, Colors } from './types'

export const getNetworkData = (
  routes: Route[],
  hunters?: string[],
  start?: string,
  finish?: string,
): Network => {
  const planets = new Set<string>()
  const links: Link[] = []

  routes.forEach(({ origin: source, destination: target, travel_time: distance }) => {
    planets.add(source)
    planets.add(target)
    links.push({ source, target, distance })
  })

  return {
    nodes: Array.from(planets).map((id) => ({
      id,
      color: mapColor(id, start, finish),
      start: id === start,
      finish: id === finish,
      hunter: (hunters || []).includes(id),
    })),
    links,
  }
}

const mapColor = (id: string, start?: string, finish?: string): Colors => {
  if (id === start) return Colors.START
  if (id === finish) return Colors.FINISH
  return Colors.DEFAULT
}
