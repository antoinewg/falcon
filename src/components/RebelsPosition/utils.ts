// Constant for now
const PLANETS = ['Tatooine', 'Dagobah', 'Hoth', 'Endor']

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const getErrorMessage = (rebels: any): string | undefined => {
  let errorMessage

  if (typeof rebels !== 'object' || !rebels) return 'The object is malformed'
  if (!('countdown' in rebels)) return "The object should have the key 'countdown'"
  if (!('bounty_hunters' in rebels)) return "The object should have the key 'bounty_hunters'"

  const hunters = 'bounty_hunters' in rebels ? rebels.bounty_hunters : []
  if (!Array.isArray(hunters)) return "'bounty_hunters' should be an array"
  if (hunters.some(({ planet }) => !PLANETS.includes(planet))) return 'Some planets are incorrect'

  return errorMessage
}
