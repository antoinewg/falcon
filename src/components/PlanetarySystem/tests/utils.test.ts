import snapshotDiff from 'snapshot-diff'

import { getNetworkData } from '../utils'

const routes = [
  {
    origin: 'Tatooine',
    destination: 'Dagobah',
    travel_time: 6,
  },
  {
    origin: 'Dagobah',
    destination: 'Endor',
    travel_time: 4,
  },
  {
    origin: 'Dagobah',
    destination: 'Hoth',
    travel_time: 1,
  },
  {
    origin: 'Hoth',
    destination: 'Endor',
    travel_time: 1,
  },
  {
    origin: 'Tatooine',
    destination: 'Hoth',
    travel_time: 6,
  },
]

const baseNetwork = {
  nodes: [
    { id: 'Tatooine', color: 'lightskyblue', finish: false, start: false, hunter: false },
    { id: 'Dagobah', color: 'lightskyblue', finish: false, start: false, hunter: false },
    { id: 'Endor', color: 'lightskyblue', finish: false, start: false, hunter: false },
    { id: 'Hoth', color: 'lightskyblue', finish: false, start: false, hunter: false },
  ],
  links: [
    { source: 'Tatooine', target: 'Dagobah', distance: 6 },
    { source: 'Dagobah', target: 'Endor', distance: 4 },
    { source: 'Dagobah', target: 'Hoth', distance: 1 },
    { source: 'Hoth', target: 'Endor', distance: 1 },
    { source: 'Tatooine', target: 'Hoth', distance: 6 },
  ],
}

describe('getNetworkData()', () => {
  it('should retrieve the network data from routes, with no hunters and no start/finish', () => {
    const network = getNetworkData(routes, [], undefined, undefined)
    expect(network.links).toStrictEqual(baseNetwork.links)
    expect(network.nodes).toStrictEqual(baseNetwork.nodes)
  })

  it('should retrieve the network data from routes, with no hunters but with start/finish', () => {
    const network = getNetworkData(routes, [], 'Tatooine', 'Dagobah')
    expect(network.links).toStrictEqual(baseNetwork.links)
    expect(snapshotDiff(network.nodes, baseNetwork.nodes)).toMatchSnapshot()
  })

  it('should retrieve the network data from routes, with no hunters but with invalid start/finish', () => {
    const network = getNetworkData(routes, [], 'Totooine', 'Dagobert')
    expect(network.links).toStrictEqual(baseNetwork.links)
    expect(network.nodes).toStrictEqual(baseNetwork.nodes)
  })

  it('should retrieve the network data from routes, with no start/finish but with hunters', () => {
    const network = getNetworkData(routes, ['Hoth'], undefined, undefined)
    expect(network.links).toStrictEqual(baseNetwork.links)
    expect(snapshotDiff(network.nodes, baseNetwork.nodes)).toMatchSnapshot()
  })

  it('should retrieve the network data from routes, with start/finish and hunters', () => {
    const network = getNetworkData(routes, ['Hoth'], 'Tatooine', 'Dagobah')
    expect(network.links).toStrictEqual(baseNetwork.links)
    expect(network.nodes).toStrictEqual([
      { id: 'Tatooine', color: 'lightsalmon', finish: false, start: true, hunter: false },
      { id: 'Dagobah', color: 'lightgreen', finish: true, start: false, hunter: false },
      { id: 'Endor', color: 'lightskyblue', finish: false, start: false, hunter: false },
      { id: 'Hoth', color: 'lightskyblue', finish: false, start: false, hunter: true },
    ])
  })
})
