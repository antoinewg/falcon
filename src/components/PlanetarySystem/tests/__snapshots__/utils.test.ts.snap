// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNetworkData() should retrieve the network data from routes, with no hunters but with start/finish 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,16 +1,16 @@
  Array [
    Object {
-     \\"color\\": \\"lightsalmon\\",
+     \\"color\\": \\"lightskyblue\\",
      \\"finish\\": false,
      \\"hunter\\": false,
      \\"id\\": \\"Tatooine\\",
-     \\"start\\": true,
+     \\"start\\": false,
    },
    Object {
-     \\"color\\": \\"lightgreen\\",
-     \\"finish\\": true,
+     \\"color\\": \\"lightskyblue\\",
+     \\"finish\\": false,
      \\"hunter\\": false,
      \\"id\\": \\"Dagobah\\",
      \\"start\\": false,
    },
    Object {"
`;

exports[`getNetworkData() should retrieve the network data from routes, with no start/finish but with hunters 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -21,10 +21,10 @@
      \\"start\\": false,
    },
    Object {
      \\"color\\": \\"lightskyblue\\",
      \\"finish\\": false,
-     \\"hunter\\": true,
+     \\"hunter\\": false,
      \\"id\\": \\"Hoth\\",
      \\"start\\": false,
    },
  ]"
`;
